Initial changes for BS 1.12.2

ErisApps opened the pull request:
This PR also fixes a NullRef crash related to the HoverHintController, but I'm honestly not sure if it works now as it should so please verify that on your end as well 🙂 


### My last commit adds a wide variety of changes across the board including (but not exhaustive as I might have forgotten some things)

- Zenject for UI. Currently excluding the custom markdown component.
- Fixed a wide variety issues related to the new curved UI.
- Finally fixed the confirmation dialog being so wack.
- Started using BSMT.T and SDK-style csproj (manifest version is now single source of truth and beatmods zips will be created when building in RELEASE mode)
- Added .editorconfig file
- Other changes that I might have missed

Currently relies on a yet-to-be-released BSML version as it fixes something in BSML that would crash the UI.
( https://github.com/monkeymanboy/BeatSaberMarkupLanguage/pull/74 )


Yet, there are still things left to do or fix...
#### Known bugs / TODOs:
- Logger injection in custom components / zenjectifying the custom components
- ~~Wack behavior when trying to disable/enable SingleStartInit mods~~
(Was supposedly fixed in https://github.com/bsmg/BeatSaber-IPA-Reloaded/commit/6b2dc6990a6d75373df648508d26e4b48a8ef468, but feel free to double-check )
- Some markdig components aren't curved correctly just yet
- Scrollbar is gone for one reason or another
- Still need to add confirmation dialog before committing the changeTransaction

ErisApps committed:
Initial changes for BS 1.12.2 update

ErisApps committed:
Fixed HoverHintController NullRef crash

ErisApps committed:
Fixed some bsml layout files not being included

ErisApps committed:
Looks ugly but somewhat fixes bsml errors/warnings

ErisApps committed:
Fixed the markdig stuff for curvature (I hope)

ErisApps committed:
A whole lot of changes across the board

- Zenject for UI. Currently excluding the custom markdown component.
- Fixed a wide variety issues related to the new curved UI.
- Finally fixed the confirmation dialog being so wack.
- Started using BSMT.T and SDK-style csproj (manifest version is now single source of truth and beatmods zips will be created when building in RELEASE mode)
- Added .editorconfig file
- Other changes that I might have missed

Currently relies on a yet-to-be-released BSML version as it fixes something in BSML that would crash the UI

Known TODOs:
- Logger injection in custom components / zenjectifying the custom components
- Wack behavior when trying to disable/enable SingleStartInit mods
- Some markdig components aren't curved correctly just yet
- Scrollbar is gone for one reason or another
- Still need to add confirmation dialog before committing the changeTransaction

ErisApps committed:
Changed .editorconfig regarding field/property/constants naming

nike4613 commented on ModList/Helpers/Helpers.cs:
```diff
@@ -0,0 +1,177 @@
+        private static Sprite? oSprite;
+        public static Sprite OSprite => oSprite ??= ReadImageFromSelf(RESOURCE_PREFIX + "o.png").AsSprite()!;
+
+        private const string WARN_SPRITE_RESOURCE_PATH = RESOURCE_PREFIX + "!.png";
```
I mentioned this in DMs, but I'm not a fan of full caps constants. I'd like to keep them as PascalCase.

ErisApps replied:
I might have missed this when adjusted the .editorconfig file. Will check.

nike4613 commented on ModList/ModList.csproj:
```diff
@@ -1,237 +1,114 @@
+    </ItemGroup>
+
+    <ItemGroup>
+        <PackageReference Include="BeatSaberModdingTools.Tasks" Version="1.3.2">
```
How well is BSMT supported without the extension? I don't use the extension, and prefer my BuildTools solution anyway. That being said, I've been meaning to make it less janky for quite a while now.

ErisApps replied:
It works perfectly fine without the extensions. I'm using Rider so BSMT isn't available for that. 😅

nike4613 commented on ModList/Services/ModProviderService.cs:
```diff
@@ -0,0 +1,42 @@
+    {
+        /// <remark>
+        /// I really didn't want to make it static...
+        /// but due to BSIPA' behaviour regarding how it handles enabling/disabling singleStartInit mods... this really is necessary...
```
I'm not sure I understand why this needs to be static. Could you explain?

ErisApps replied:
Explained in DMs. Will need more investigation though.

nike4613 commented on ModList/UI/BSML/MarkdownTextHandler.cs:
```diff
@@ -18,21 +15,46 @@ public class MarkdownTextHandler : TypeHandler<MarkdownText>
+                {
+                    "autolink-color"
+                }
+            }
```
These do not need to be exploded like they are. It is so much more readable when each one is on its own line.

ErisApps replied:
Will look into the .editorconfig file for this.

nike4613 commented on ModList/UI/BSML/MarkdownTextHandler.cs:
```diff
@@ -18,21 +15,46 @@ public class MarkdownTextHandler : TypeHandler<MarkdownText>
-                { "linkColor", (md, col) => md.LinkColor = GetColor(col, md.LinkColor) },
-                { "autolinkColor", (md, col) => md.AutolinkColor = GetColor(col, md.AutolinkColor) },
-            };
+        public override Dictionary<string, Action<MarkdownText, string>> Setters { get; } = new Dictionary<string, Action<MarkdownText, string>>
```
This can use C# 9 and just be `new() ...`

ErisApps replied:
C# 9 is sadly not allowed for BeatMods mods for the time being as per request of @Auros

nike4613 replied:
That seems... odd. Did he give a reason?

Auros replied:
I requested for mods to not use C# 9 since dnSpy can't fully decompile some C# 9 code without it looking funky, and other popular deobfuscators that are used by approvers like JustAssembly will just straight up crash.

If I were the only BeatMods approver i probably wouldn't have an issue and go the extra mile to verify stuff, but I'm not.

nike4613 replied:
IIRC the only things C# 9 thing that dnSpy *really* doesn't like is function pointers. Most stuff works fine.

ErisApps replied:
In the meantime, I wrote a approval helper application based on the ILSpy.Decompiler package which can automatically diff between mod versions so C# 9 _should_ be fine to use now if you really feel like it.

nike4613 commented on ModList/UI/Components/MarkdownText.cs:
```diff
@@ -101,26 +101,31 @@ public UnityRenderer Renderer
-                    var ufont = new Font(config.MonospaceFontPath);
-                    asset = Helpers.TMPFontFromUnityFont(ufont);
+                    var uFont = new Font(config.MonospaceFontPath);
+                    asset = Helpers.Helpers.TMPFontFromUnityFont(uFont);
```
Is there a reason Helpers needed to be in its own namespace of the same name? Generally I'd do `Utilities.Helpers` to avoid this redundancy, or just have it be in the root namespace if its only the one class.

nike4613 commented on ModList/UI/Components/MarkdownText.cs:
```diff
@@ -101,26 +101,31 @@ public UnityRenderer Renderer
-                if (!FontManager.TryGetTMPFontByFullName(config.MonospaceFontName, out asset, setupOsFallbacks: false)
-                 && !FontManager.TryGetTMPFontByFamily(config.MonospaceFontName, out asset, setupOsFallbacks: false))
+                if (!FontManager.TryGetTMPFontByFullName(config.MonospaceFontName, out asset, setupOsFallbacks: false) &&
+                    !FontManager.TryGetTMPFontByFamily(config.MonospaceFontName, out asset, setupOsFallbacks: false))
```
I tend to prefer to have the operator at the beginning of the next line rather than the end of the last. Makes it easier to tell at a glance what the condition is.

nike4613 commented on ModList/UI/Components/MarkdownText.cs:
```diff
@@ -283,20 +294,20 @@ public void OnPointerEnter(PointerEventData eventData)
             {
                 get => _hoverHintController;
-                set => _hoverHintController = value;
+                set => this.SetField<HoverHint, HoverHintController>("_hoverHintController", value);
```
I take it the field was made readonly? Do we need to actually set the field most of the time?

Also, `nameof`, please.

ErisApps replied:
It was indeed marked as readonly, hence the use of  `.SetField<T, U>()`.
That being said, I simply fixed this as it prevented the code from compiling, but I honestly don't know whether this is actually being used at all...

Forgot to use `nameof ()` so will fix that.

nike4613 commented on ModList/UI/Components/TextMeshProUGUILinkBackgroundGenerator.cs:
```diff
@@ -22,24 +21,26 @@ public class TextMeshProUGUILinkBackgroundGenerator : MonoBehaviour
             {
                 useLineHeight = value;
-                needsRerender = true;
+                _needsRerender = true;
```
I also mentioned this in DMs; I really dislike the `_` prefix for private fields.

nike4613 commented on ModList/UI/Markdig/RenderHelpers.cs:
```diff
@@ -7,57 +7,79 @@ namespace IPA.ModList.BeatSaber.UI.Markdig
+                2 => EmphasisFlags.Bold,
+                var i when i > 2 => EmphasisFlags.Italic | EmphasisFlags.Bold,
                 _ => EmphasisFlags.None
-            };
```
For stuff like this, its more obvious to me that its an expression switch if the body is indented one more level and the expression body arrow is on the line below the declaration.

nike4613 commented on ModList/UI/Markdig/RenderHelpers.cs:
```diff
@@ -7,57 +7,79 @@ namespace IPA.ModList.BeatSaber.UI.Markdig
-            => builder.AppendFormat("{0:X2}{1:X2}{2:X2}{3:X2}",
-                (int)(color.r * 255), (int)(color.g * 255), 
-                (int)(color.b * 255), (int)(color.a * 255));
+        public static StringBuilder AppendColorHex(this StringBuilder builder, Color color) => builder.AppendFormat("{0:X2}{1:X2}{2:X2}{3:X2}",
```
Much the same as above; I prefer the expression bodies, if they aren't trivial, to be on the next line and everything else more indented.

nike4613 commented on ModList/UI/Markdig/UnityRenderer.cs:
```diff
@@ -58,19 +59,26 @@ public class UnityRenderer : IMarkdownRenderer
-                Inline inline => RenderInline(inline, ParagraphFontSize),
-                _ => throw new NotImplementedException("Unknown markdown object type")
-            };
+        object IMarkdownRenderer.Render(MarkdownObject obj) => obj switch
```
See previous files.

nike4613 commented on ModList/UI/Markdig/UnityRenderer.cs:
```diff
@@ -159,57 +171,68 @@ private IEnumerable<RectTransform> RenderParagraph(ParagraphBlock para)
             AfterObjectRendered?.Invoke(para, transform.gameObject);
 
-            return Helpers.SingleEnumerable(transform).Append(Spacer(1.5f));
+            return transform.SingleEnumerable().Append(Spacer(1.5f));
```
The reason I was using this method explicitly rather than as an extension was because I realized that it was more readable not as an extension method. I must have forgotten to make it not an extension method.

nike4613 commented on ModList/UI/Markdig/UnityRenderer.cs:
```diff
@@ -335,19 +363,18 @@ private IEnumerable<RectTransform> RenderListItem(ListItemBlock item, bool isLoo
-                        children = RenderInline(para.Inline, ParagraphFontSize);
-                    else
-                        children = RenderBlock(block);
+                    children = block is ParagraphBlock {Inline: { }} para ? RenderInline(para.Inline, ParagraphFontSize) : RenderBlock(block);
```
Spaces inside braces please :)

ErisApps replied:
.editorconfig has been adjusted for that 😄

nike4613 commented on ModList/UI/Markdig/UnityRenderer.cs:
```diff
@@ -389,92 +419,91 @@ private IEnumerable<RectTransform> RenderInline(Inline inline, float fontSize, b
+                .Append("</noparse>")
+                .Append(InlineCodePaddingText)
+                .Append("</link></size>")
+                .Append(CodeFont == null ? "" : "</font>");
```
I typically like having all of the chained method calls aligned with the first, though I don't care about that quite as much as some of the other things I've mentioned.

nike4613 commented on ModList/UI/Markdig/UnityRenderer.cs:
```diff
@@ -495,31 +524,32 @@ private StringBuilder RenderAutolinkInlineToText(AutolinkInline link, StringBuil
         private struct LinkInfo
         {
-            public string Url;
+            public string URL;
```
If the BCL class is `Uri`, this is `Url`.

nike4613 commented on ModList/UI/Markdig/UnityRendererBuilder.cs:
```diff
@@ -66,7 +66,7 @@ public interface IInlineCodeRendererBuilder
         }
 
-        public interface IUIRendererBuidler
+        public interface IuiRendererBuidler
```
What on earth is this capitalization?

ErisApps replied:
I... erm... might have messed up here...

nike4613 commented on ModList/UI/ModListFlowCoordinator.cs:
```diff
@@ -1,87 +1,120 @@
 namespace IPA.ModList.BeatSaber.UI
 {
-    internal class ModListFlowCoordinator : FlowCoordinator
+    public class ModListFlowCoordinator : FlowCoordinator
```
Why should the flow coordinator be public?

ErisApps replied:
Accidental change

nike4613 commented on ModList/UI/ModListFlowCoordinator.cs:
```diff
@@ -1,87 +1,120 @@
-            base.BackButtonWasPressed(topViewController);
-            ParentFlowCoordinator.DismissFlowCoordinator(this);
+            // Check whether there's a transaction going on and commit :eyes:
+            // TODO: Guess it's a good idea to add a confirmation here as well...
```
It would probably also be worth adding an additional confirmation if they're disabling ModList.

ErisApps replied:
Can I just add it to the TODO or did I have to actually implement this as well before the PR can be merged?

nike4613 replied:
As a TODO is probably fine

nike4613 commented on ModList/UI/ViewControllers/ModControlsViewController.cs:
```diff
@@ -1,21 +1,38 @@
-    [HotReload(PathMap = new[] { "C:\\", CompileConstants.SolutionDirectory })]
-    public class ModControlsViewController : BSMLAutomaticViewController
+    [HotReload(RelativePathToLayout = @"..\Views\ModControlsView.bsml")]
+    [ViewDefinition("IPA.ModList.BeatSaber.UI.Views.ModControlsView.bsml")]
```
I generally prefer having the BSML next to the ViewController exactly so that this attribute isn't necessary, as well as generally liking those things to go together since their inextricably tied to each other anyway.

nike4613 commented on ModList/UI/ViewControllers/ModControlsViewController.cs:
```diff
@@ -252,29 +262,43 @@ private Action<bool> GetDisableConfirmCallback(PluginInformation plugin, IEnumer
-            {
-                yield return plugin.Name;
-            }
+            return plugins.Select(plugin => plugin.Name);
```
This should be able to be an expression body method.

nike4613 commented on ModList/UI/ViewControllers/ModalPopupViewController.cs:
```diff
@@ -0,0 +1,158 @@
+        internal bool HasCurrentChangeLines => CurrentChange?.LineEntries.Any() ?? false;
+
+        [UIValue("current-change-lines")]
+        internal string CurrentChangeLines => CurrentChange?.LineEntries.Aggregate("<color=\"yellow\">", (accumulator, modName) => $"{accumulator}{modName}\n") ?? string.Empty;
```
This is long enough that I'd want the body on the next line.

ErisApps committed:
Removed version binding and fixed name binding

ErisApps committed:
Fixed some access modifiers

ErisApps committed:
Fixed naming in Helpers.cs

ErisApps committed:
Moved Helpers.cs to Utils namespace to prevent Helpers.Helpers in code

ErisApps committed:
Use nameof in SetField

ErisApps committed:
Fixed more naming and moved logical operation to beginning of nextline

ErisApps committed:
Reformatted stuff

Editorconfig might override it again, but still need to look into changing that eventually

ErisApps committed:
Made SingleEnurable a non-extension method

ErisApps committed:
Bump SiraUtil minimum version to 2.3.1 in manifest

ErisApps committed:
Fixed more naming

ErisApps committed:
Expression body changes

ErisApps committed:
Misremembered the suggested Helpers namespace. it's fixed now though.

ErisApps committed:
Move views into same folder as viewcontrollers

ErisApps committed:
Even more formatting and naming

Adjust editorconfig file for spaces right after and before curly braces

ErisApps committed:
Align daisy-chained method calls

ErisApps committed:
Fixed access modifiers for installers + changed AppInstaller config init

ErisApps committed:
Fixed incorrect namespace for modal popup view

ErisApps committed:
Bumped BSML version to 1.4.3 in manifest

monkeymanboy/BeatSaberMarkupLanguage#74 is available since that version

ErisApps committed:
Bumped dependency versions

ErisApps committed:
Code cleanup

Fixed SiraUtil flowcoordinator binding deprecation

ErisApps committed:
No more static pluginList in ModProviderService

ErisApps committed:
Restructured manifest
